### PR TITLE
Import and export Sandcastle examples using Gists

### LINKED ISSUE #3702: Ability to import code samples into Sandcastle
Perhaps using

* (shortened) encoded urls (for import and export), or
* GitHub gists

as discussed offline with @mramato.

Does anyone want to mentor @TomPed on this?

### PR DESCRIPTION
**Pull request by TomPed**
Fixes #3702, #2152

Things to do:
* [x] Log errors to the sandcastle console when importing gists
* [x] Fix share. Right now when you click share you get the id of a gist but I want this to be a link that will open up sandcastle and paste the code in from the gist with that id
* [x] Fix icons
* [x] Fix textbox

### CODE REVIEW

**Comment by pjcozzi** [comment]
* Update CHANGES.md
* Do you need icons for import/export?

**Comment by pjcozzi** [comment]
@TomPed please review #2152.  We may be able to close that and #3702 with this PR.

**Comment by TomPed** [comment]
These are the icons that I am using right now:

screen shot 2016-04-01 at 9 58 07 am

What do you think?

**Comment by pjcozzi** [comment]
Can we use the standard share icon?  @ggetz can advise on what we used for the website.

And perhaps use the GitHub icon for import gist if their terms of use allow it.

**Comment by ggetz** [comment]
Share icons are usually represented by http://fortawesome.github.io/Font-Awesome/icon/share-square-o/ or http://fortawesome.github.io/Font-Awesome/icon/share-alt/

**Comment by TomPed** [comment]
Okay I will try to find a `dijit` icon similar since that seems to be the theme of sandcastle.

**Comment by TomPed** [comment]
Still a few things to do, listed up top.

**Comment by TomPed** [comment]
Added icons. Need to fix textbox issue.

**Comment by TomPed** [comment]
I've made some changes. All I need to do now is to not auto run code on page load. If you click import gist the code does not auto run, but if you enter a url with the query parameter for a gist the code will run automatically, which could cause security issues. I'm trying to figure out a fix for this.

**Comment by TomPed** [comment]
Fixed auto run problem. @pjcozzi one last thing I think. We discussed having a header when we share an example in the GitHub Gist (like the Cesium version and stuff like that). What if someone imports a Gist, with the header. Tweaks the code and then creates a new gist. That gist will have a double header. Is it worth fixing this (with an easy if) or should we not include the header? If we want the header what should it include besides the Cesium version?

**Comment by TomPed** [comment]
History works for gist, but not for other sandcastle examples. Working on that fix now.

File: ThirdParty/dojo-release-1.10.4/dijit/icons/commonIcons_rtl.css [hunk 1]
```diff
@@ -119,3 +119,17 @@ The 16 x 16px icons in these sprites are action and object type images which can
 .dijitRtl .dijitDisabled .dijitIconError {
 	background-image: url('images/commonIconsObjActDisabled_rtl.png'); /* Contains both object and action icons as a sprite image for the disabled state. These would be used by buttons and menus.   */
 }
+
+.gitHubIcon {
```
Comment (pjcozzi) [inline ThirdParty/dojo-release-1.10.4/dijit/icons/commonIcons_rtl.css]: This file is part of Dojo, which is a third-party library.  Do not modify it.  There's probably a .css file in the `Apps/Sandcastle` directory you can use.  If not, @emackey will know.

**Comment by pjcozzi** [comment]
Do we need to update https://github.com/AnalyticalGraphicsInc/cesium/blob/master/LICENSE.md#example-applications for either of the new icons?

**Comment by pjcozzi** [comment]
Merge in master.  There is probably a trivial conflict in CHANGES.md.

**Comment by pjcozzi** [comment]
> History works for gists, but not for other sandcastle examples. Working on that fix now.

I don't know what you mean.

**Comment by pjcozzi** [comment]
> We discussed having a header when we share an example in the GitHub Gist (like the Cesium version and stuff like that). What if someone imports a Gist, with the header. Tweaks the code and then creates a new gist. That gist will have a double header. Is it worth fixing this (with an easy if) or should we not include the header? If we want the header what should it include besides the Cesium version?

It would be OK to have a running history log, but the gist is already backed by a git repo.  I was thinking to include the Cesium version and save date/time, which could be useful if there is a breaking API change in Cesium.

However, why not start with no header at all, and then let's see if we actually need it.

**Comment by TomPed** [comment]
I am looking into if we need to updated https://github.com/AnalyticalGraphicsInc/cesium/blob/master/LICENSE.md#example-applications. And do so if needed.  

> I don't know what you mean.

History is not working correctly. This has been an issue for a little while now. I think the best thing to do would be to create a new HTML file and have the link to the gist open that. I will get this change out before lunch so you can review.

**Comment by TomPed** [comment]
From the font awesome website.

> Attribution is no longer required as of Font Awesome 3.0 but is much appreciated: "Font Awesome by Dave Gandy - http://fontawesome.io".

As for the GitHub icon I don't think we need to document that either. From the GitHub page.

> Use the Octocat or GitHub logo to advertise that your product has built-in GitHub integration

**Comment by pjcozzi** [comment]
Let's still include them just for attribution.

**Comment by TomPed** [comment]
I have browser history with gist some what working. I will be talking with @emackey tomorrow to discuss my approach and help me out.

**Comment by TomPed** [comment]
I now believe bowser history is 100% working. I think this is ready for a look!

**Comment by pjcozzi** [comment]
@emackey @shunter do you want to look at this at all?

**Comment by TomPed** [comment]
As discussed offline with @emackey. Right now when you navigate to a exported sandcastle example the code does not auto run for security reasons. However since cesiumjs.org does not hold any user account info we think that we can allow the code from the gist to auto run. Does anyone anyone else see a security issue here?

**Comment by mramato** [comment]
I can't think of a good reason not to auto-run the code.

Also, I discussed with @TomPed a bit offline, but I think the share button should probably pop up a modal with the link filed in rather than have that weird empty text field it has right now.  The modal could also let the user know that further edits to the code will not be reflected in the link (which is non-obvious right now).

**Comment by TomPed** [comment]
I've revamped the importing. Check it out.

**Comment by pjcozzi** [comment]
If there are no security concerns, auto-run is OK with me.

@TomPed is there anything else before we merge this?

Does anyone else want to review this?

**Comment by TomPed** [comment]
I will make that change now. I will also make some changes with the error handling. And do one final "user test".

**Comment by TomPed** [comment]
Made the changes that I wanted. If anyone else would take a look, that'd be great!

**Comment by pjcozzi** [comment]
Can the "Share" dialog appear right under the toolbar button instead of in the center of the page?

Have you tried the share workflow for a few different apps?  I know we discussed not wanting to create the GitHub gist when they first click the toolbar button, but this feels awkward to me.  For example, shouldn't the "Get Link" button grey or go away once the link is shown?  Should we auto select-all?  Should just clicking in the text box create the link?  What is simple and painful for the user?

**Comment by pjcozzi** [comment]
Fun idea: could be cool to write a Chrome plugin that can open a GitHub gist in Sandcastle.  Total overkill right now.

**Comment by pjcozzi** [comment]
Let's not add it yet, but I wonder if long-term, we'll want a link to the GitHub gist from Sandcastle, e.g., https://gist.github.com/anonymous/7d67199111d68c6a6eceef8d0369192e, for comments and revision history.

**Comment by pjcozzi** [comment]
Did you test sharing when your offline?  Is there a reasonable error message?

**Comment by pjcozzi** [comment]
Works great!  Just those comments.

**Comment by TomPed** [comment]
Okay so I've changed how sharing works. I think that this is very easy to use and similar to how a few websites share. Also you will notice if you don't change code after sharing and try to share again the same link will be in the textbox. @pjcozzi what do you think?

**Comment by TomPed** [comment]
Made error messages reasonable when offline.

**Comment by pjcozzi** [comment]
This is all good with me.

Does anyone else want to review before I merge?

**Comment by pjcozzi** [comment]
Thanks again @TomPed!

**Pull request merged**